src/__tests__/poller.test.ts

```diff
@@ -0,0 +1,135 @@
+import Poller from '../poller'
+import type { Options } from '../types'
+import { getRandom } from '../utils'
+
+jest.useFakeTimers()
+
+jest.mock('../utils.ts', () => ({
+  getRandom: jest.fn()
+}))
+
+interface PollerArgs {
+  fetchFlags: jest.MockedFunction<() => Promise<any>>
+  configurations: Partial<Options>
+  pollInterval: number
+  maxAttempts: number
+}
+
+interface TestArgs {
+  delayFunction: jest.MockedFunction<() => number>
+  delayMs: number
+}
+
+const getPoller = (overrides: Partial<PollerArgs> = {}): { poller: Poller; pollerArgs: PollerArgs } => {
+  const args: PollerArgs = {
+    fetchFlags: jest.fn(),
+    configurations: {},
+    pollInterval: 60000,
+    maxAttempts: 5,
+    ...overrides
+  }
+
+  return {
+    poller: new Poller(args.fetchFlags, args.configurations, args.pollInterval),
+    pollerArgs: args
+  }
+}
+
+const getTestArgs = (overrides: Partial<TestArgs> = {}): TestArgs => {
+  const { delayMs = 5000 } = overrides // Extract delayMs or set a default of 5000
+
+  return {
+    delayMs,
+    delayFunction: (getRandom as jest.Mock).mockReturnValue(delayMs),
+    ...overrides
+  }
+}
+
+describe('Poller', () => {
+  it('should not start polling if it is already polling', () => {
+    const { poller } = getPoller()
+    const logSpy = jest.spyOn(poller as any, 'logDebug')
+
+    poller.start()
+    poller.start()
+
+    expect(logSpy).toHaveBeenCalledWith(
+      'Polling was requested but is already running - only one poller can run at a time.'
+    )
+  })
+
+  it('should retry fetching if there is an error', async () => {
+    const { poller, pollerArgs } = getPoller()
+    const fetchFlags = pollerArgs.fetchFlags
+    fetchFlags.mockResolvedValue(new Error('Fetch Error'))
+
+    const testArgs = getTestArgs()
+
+
+    const logSpy = jest.spyOn(poller as any, 'logDebug')
+
+    poller.start()
+
+    jest.advanceTimersByTime(pollerArgs.pollInterval)
+
+    // Allow first attempt to resolve
+    await Promise.resolve()
+
+    // Advance past the first delay
+    jest.advanceTimersByTime(testArgs.delayMs)
+
+    // Allow successful attempt to resolve
+    await Promise.resolve()
+
+    expect(fetchFlags).toHaveBeenCalledTimes(2)
+    expect(logSpy).toHaveBeenCalledWith('Error when polling for flag updates', expect.any(Error))
+  })
+
+  it('should not retry after max attempts are exceeded', async () => {
+    const { poller, pollerArgs } = getPoller()
+    const fetchFlags = pollerArgs.fetchFlags
+    const pollInterval = pollerArgs.pollInterval
+    const maxAttempts = pollerArgs.maxAttempts
+    fetchFlags.mockImplementation(() => {
+      return Promise.resolve(new Error('Fetch flags Error'))
+    })
+
+    const testArgs = getTestArgs()
+
+    const logSpy = jest.spyOn(poller as any, 'logDebug')
```

Comment:
Again, do not test implementation details. Instead observe what the user can observe. In this case you can set configurations as `{ debug: true }` and then spying on `console.debug`

Reply:
Thanks - tests updated

Reply:
I've actually removed the assertions for checking if console.debug was called. No matter what I did, the Spy's state for that was bleeding over into each test. It was really strange. Burning time trying to fix it for next to no value so removed it.

Reply:
```
beforeEach(() => {
  jest.clearAllMocks()
})
```

Reply:
I tried that - it really wasn't working. Was the strangest thing. By the time the last test ran, console.debug was called 25 times!

Reply:
You're gonna absolutely kick yourself. The issue was that you never stopped the previous pollers when their tests were done, so they kept going in the background! The fix is kinda silly, but easy.

```typescript
let currentPoller: Poller

const getPoller = (overrides: Partial<PollerArgs> = {}): Poller => {
  const args: PollerArgs = {
    fetchFlags: jest.fn(),
    configurations: {},
    ...overrides
  }

  currentPoller = new Poller(args.fetchFlags, args.configurations)

  return currentPoller
}

// ... then down in your describe block

afterEach(() => {
  jest.clearAllMocks()
  currentPoller?.stop()
})
```